Add Step and Sidecar Workspaces feature

sbwsg opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

TEP: https://github.com/tektoncd/community/blob/master/teps/0029-step-workspaces.md
Fixes https://github.com/tektoncd/pipeline/issues/2343

Access to Workspaces can now be limited to specific Steps and Sidecars in a Task.
This allows Task authors to isolate sensitive data to specific images, reducing
exposure of assets like credentials.

As a side-effect of this change Workspaces are now mounted to Sidecars by
default, just as they are to Steps.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
- New alpha feature added: "Step and Sidecar Workspaces"
    - Requires "enable-api-fields: alpha" feature gate.
    - Sidecars in a Task now receive Workspaces, just like Steps do.
    - Steps and Sidecars can now specify which Workspaces they want access to. This isolates the workspace to only those Steps or Sidecars that need them, allowing Task authors to limit exposure of sensitive data like credentials.
    - The default behaviour for workspaces in Steps remains the same - all Steps get access to them.
```

sbwsg commented:
Build tests are failing with following message:

```
/home/prow/go/src/github.com/tektoncd/pipeline is out of date. Please run hack/update-codegen.sh
```

But `update-codegen.sh` doesn't create any new code when I run it.  This error is actually appearing because `update-openapigen.sh` needs to be run.  I'm going to update that error message as part of this PR.

bobcatfish reviewed (commented):
Started with some initial comments on docs and examples, all super minor so far! Will return to review more ASAP

sbwsg commented on docs/workspaces.md:
```diff
@@ -138,38 +145,47 @@ spec:
-**Note:** `Sidecars` _must_ explicitly opt-in to receiving the `Workspace` volume. Injected `Sidecars` from
-non-Tekton sources will not receive access to `Workspaces`.
+It can potentially be useful to mount `Workspaces` to different locations on a per-`Step` or
+per-`Sidecar` basis and this is also supported:
```
Yeah, exactly.  The moment a `workspace` entry is present on a step/sidecar it's considered isolated.

sbwsg commented on docs/workspaces.md:
```diff
@@ -88,6 +88,13 @@ many uses:
+Certain kinds of data are more sensitive than others. To reduce exposure of sensitive data Task
+authors can isolate `Workspaces` to only those `Steps` and `Sidecars` that require access to
+them. The primary use-case for this is credentials but it can apply to any data that should have
+its access strictly limited to only specific container images.
```
Thanks for reviewing, added the link.

sbwsg commented on examples/v1beta1/taskruns/workspace-isolation.yaml:
```diff
@@ -0,0 +1,50 @@
+metadata:
+  generateName: workspace-isolation-
+spec:
+  timeout: 2m0s
```
Updated with comments in both loops mentioning the use of `timeout` to prevent looping forever.

sbwsg commented on hack/verify-codegen.sh:
```diff
@@ -52,6 +52,6 @@ if [[ $ret -eq 0 ]]
   echo "${REPO_ROOT_DIR} up to date."
 else
-  echo "${REPO_ROOT_DIR} is out of date. Please run hack/update-codegen.sh"
+  echo "${REPO_ROOT_DIR} is out of date. Please run hack/update-codegen.sh and hack/update-openapigen.sh"
```
Good catch, I've updated DEVELOPMENT.md.

bobcatfish commented on pkg/workspace/apply_test.go:
```diff
@@ -522,3 +722,23 @@ func TestApply(t *testing.T) {
+	}}
+	vols := workspace.CreateVolumes(bindings)
+	_, err := workspace.Apply(ts, bindings, vols)
+	if err == nil || !strings.Contains(err.Error(), "did not exist in declared workspaces") {
```
nit: i find that asserting against error messages (or log messages) is a bit brittle to maintain over time - if it's important to assert that the correct error is returned, i usually go to one of 2 alternatives:

* return different types of errors  (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully is a very interesting post on error handling in Go and iirc doenst like that approach either)
* treat the different error types as a sign that one function is doing too much, i.e. if it can fail in 2 different ways and its important to distinguish between both, maybe you're better off with two separate functions (it does SEEM in this case like maybe Apply is also doing validation)

sbwsg replied:
Removed string error message check.

bobcatfish commented on pkg/workspace/apply_test.go:
```diff
@@ -509,6 +510,205 @@ func TestApply(t *testing.T) {
+			Workspaces: []v1beta1.WorkspaceDeclaration{{
+				Name: "source",
+			}},
+		},
```
nice, clear test cases!

bobcatfish commented on pkg/workspace/apply.go:
```diff
@@ -121,3 +140,45 @@ func Apply(ts v1beta1.TaskSpec, wb []v1beta1.WorkspaceBinding, v map[string]core
+			if workspaceUsage.Name == workspaceName {
+				vm := volumeMount
+				if workspaceUsage.MountPath != "" {
+					vm.MountPath = workspaceUsage.MountPath
```
it seems like this is the code that handles the mountPath overriding logic, but i dont see it being explicitly included in the tests cases below? (apologies if i just missed it) i could see there being an extra case or two for this?

bobcatfish replied:
i think the coverage report might be backing this up as well:

![image](https://user-images.githubusercontent.com/432502/106949911-a0786000-66fb-11eb-99e6-833bed7c9acf.png)

sbwsg replied:
Unit test added to cover custom mountPath in sidecar and step.

bobcatfish commented on pkg/reconciler/taskrun/resources/apply_test.go:
```diff
@@ -786,6 +786,56 @@ func TestApplyWorkspaces(t *testing.T) {
+				Name:      "ws",
+				MountPath: "/bar",
+			}},
+		}}},
```
nice!!! 😎

bobcatfish commented on pkg/reconciler/taskrun/resources/apply.go:
```diff
@@ -144,6 +144,40 @@ func ApplyWorkspaces(spec *v1beta1.TaskSpec, declarations []v1beta1.WorkspaceDec
+				v1beta1.ApplySidecarReplacements(sidecar, stringReplacements, emptyArrayReplacements)
+			}
+		}
+	}
```
just nitpicking and i dont have a clear alternative suggestion, but it feels like some of this logic belongs to how "workspaces" work and i'd almost expect to see it in the workspaces package

(i.e. which path to use feels like it's more about how workspaces work, and less about how string substitution works)

sbwsg replied:
I'm not sure there's a good way to disentangle the choice of mountpath for each case (step / sidecar / shared) from the string substitution logic.  Particularly since the variables being replaced are order-dependent (we have to replace the isolated workspace mountpaths before those that are shared). I can't see an approach that doesn't either leak some workspace logic into the string substitution package or leak some string substitution logic into the workspace package.

We can always revisit?  There's nothing user-facing here that will prevent a refactor if we figure out a better separation.

bobcatfish reviewed (commented):
I think pretty much all my comments are just idle thoughts or unimportant nitpicks! The only at all "significant" comment I have is a think there might be a bit of missing test coverage on the mount path overriding.

Looks great!

/approve

i wonder if we want to explicitly do anything about how this impacts pipelineresources? since this was implemented before using steptemplates, removing the volumes from the step template would also mean removing it from pipelineresources, but right now we don't cover it with any tests or mention it in docs - i think its probably better this way; its hard to imagine any pipelineresources that would be impacted, probably no one even knew that was happening, and lastly we probably want to discourage using pipelineresources anyway, but i wanted to bring it up and see what you think

sbwsg commented on pkg/apis/pipeline/v1beta1/workspace_types.go:
```diff
@@ -111,3 +111,12 @@ type WorkspacePipelineTaskBinding struct {
+type WorkspaceUsage struct {
+	// Name is the name of the workspace this Step or Sidecar wants access to.
+	Name string `json:"name"`
+	// MountPath is the path that the workspace should be mounted to inside the Step or Sidecar.
```
Updated comment.

sbwsg commented on docs/workspaces.md:
```diff
@@ -137,38 +150,47 @@ spec:
+  steps:
+  - workspaces:
+    - name: source-code
+      mountPath: /source
```
Updated to the more complex example.

bobcatfish commented:
![image](https://user-images.githubusercontent.com/432502/106954585-036cf580-6702-11eb-88b2-88fb28fcf87a.png)

yaaaasssss

![image](https://user-images.githubusercontent.com/432502/106954644-18e21f80-6702-11eb-94e4-8b13c489a3fb.png)

sbwsg commented:
Thanks for reviewing @bobcatfish !

pritidesai commented on docs/workspaces.md:
````diff
@@ -137,38 +150,53 @@ spec:
+    workspaces:
+    - name: foo
+      mountPath: /files # overrides mountPath
+```
````
the workspace `foo` in each step is referring to the same workspace `ws` specified under `workspaces`? should they be named same as defined under `workspaces`?

pritidesai replied:
What does "override" mean here? the step `edit-files-1` is executed with the mount path `/foo` and `edit-files-2` executed with the default mountPath `/ws`. But the files created in `edit-files-1` under `/foo` Will be visible in `edit-files-2` under `/ws`. Its just the path name which differs. Am I getting it right?

pritidesai replied:
I ran a pipeline with creating files (`/foo/sample1` and `/foo/sample2`) in `edit-files-1`. Both files were accessible as `/ws/sample1` and `/ws/sample2` in `edit-files-2`.

I don't see any issues with this behavior but may be a one liner explanation might help in the doc. Not a blocker though. 🙏

sbwsg replied:
They should be `name`d "ws".  That's a bug, well spotted! will fix.  Also I will make this a validation error if it isn't already.  "foo" isn't a valid workspace name here.

pritidesai commented on examples/v1beta1/taskruns/workspace-isolation.yaml:
```diff
@@ -0,0 +1,52 @@
+    workspaces:
+    - name: signals
+    steps:
+    - image: alpine:3.12.0
```
NIT: step names are missing

afrittoli commented:
Thank you for this, it's a great feature to add. I skimmed through the code and it looks good (thanks for nice docs 🙏 ) - the only concern I have is the lack of a feature flag. Since this adds to a beta API, adding it without a flag means that the new API field becomes beta the moment it's released.

I'm aware that we don't have a clear policy right now, I'll try to progress on https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md so that we may have better guidance for the future. 

I agree with @bobcatfish proposal to have a single API flag for (most) API changes would make sense, and I know @vdemeester expressed positive feedback to that idea, so perhaps we should add the flag along with documentation so that people may start using it. 

@sbwsg @bobcatfish @vdemeester we're almost at release time, I see two alternatives:
1. complete the work on the TEP, implement the changes associated and then merge this PR, hopefully in the next release
2. implement an API feature flag now, rebase this one on top and they might both be merged in time for this release

afrittoli commented:
/hold until the need for a feature flag is discussed / clarified

vdemeester commented:
/hold

sbwsg commented:
The feature can wait until a single shared feature flag or gate is implemented.  I will rebase on top of that work when it's in and put this behind it.

vdemeester reviewed (approved):
/meow

sbwsg commented:
I've moved this PR back to 0.24.  It's unlikely I'll be able to get this set up behind a feature gate prior to the 0.23 release early next week but no biggie if it only waits another month.

sbwsg commented:
TODO:

- [ ] Put behind alpha feature gate
- [ ] Add testing to ensure that the feature gate does indeed turn this on / off

sbwsg commented:
@bobcatfish @vdemeester @afrittoli 

I've implemented the `enable-api-fields` feature flag as part of this PR and put the new workspace feature behind the `alpha` flag.

I've moved the example YAMLs for this feature into their own directory and added some code to our examples test runner so that they only execute if the `enable-api-fields` feature flag is `alpha`.

sbwsg commented:
I'm not sure how to conditionally allow validations based on the `enable-api-fields` flag though.  I'm not sure if the validation webhook has access to the feature flag configmap.

sbwsg commented:
/test pull-tekton-pipeline-build-tests

sbwsg commented:
@vdemeester I've updated this PR to add "enable-api-fields" checks to the Validating webhook and disallow use of the new feature unless the gate is "alpha".

I'll port over the shared unit test code for checking feature gate to the Enable API Fields PR.

vdemeester commented:
> I'm not sure how to conditionally allow validations based on the `enable-api-fields` flag though. I'm not sure if the validation webhook has access to the feature flag configmap.

It should have it (we use the mechanism for default, custom-task, …).

afrittoli commented:
@sbwsg now that the feature gates PR is merged, this will need a rebase 🙏

sbwsg commented:
> now that the feature gates PR is merged, this will need a rebase 🙏

Rebased on top of new feature gates code.  I think this is ready for another look.

sbwsg commented:
I'm going to cancel the hold.  I can't find the comment where it was added and everythoing's up to date now with the feature gate work.

/hold cancel

pritidesai commented on examples/v1beta1/taskruns/workspace-in-sidecar.yaml:
```diff
@@ -19,6 +19,10 @@ spec:
     - name: signals
     steps:
     - image: alpine:3.12.0
+      resources:
```
@sbwsg I am wondering why these changes are showing up here 🤔  These were already a part of previous [PR](https://github.com/tektoncd/pipeline/commit/565c55b30e422fc7fca42bc122561aa231106204#diff-d3a6963f568541460e24acddf2bcc5d3337bbbc8632e91a63ee59c6084891a2eR22-R25) you created.

pritidesai replied:
and which was merged ...

pritidesai commented on pkg/apis/config/feature_flags.go:
```diff
@@ -118,7 +118,7 @@ func NewFeatureFlagsFromMap(cfgMap map[string]string) (*FeatureFlags, error) {
 
 // setEnabledAPIFields sets the "enable-api-fields" flag based on the content of a given map.
-// If the feature gate is invalid or missing then an error is returned.
+// If the feature gate is invalid or missing then the flag is set to its default.
```
NIT: this is not true, it does return error if the value is anything outside of `stable` and `alpha`, some git weirdness?

sbwsg replied:
Ah 😅 result of a bad merge. Good catch, thank you!

pritidesai commented on docs/workspaces.md:
```diff
@@ -135,6 +158,9 @@ spec:
 
 #### Sharing `Workspaces` with `Sidecars`
```
NIT: reading for the first time looks like, the two notes in the same section is contradicting each other. The first note is saying with `alpha` API, the workspaces are automatically available to sidecars  and second note is about `stable` API. With `stable` API, sidecars have to opt-in. This is fine for now, no need to change anything here. But we might have to repeat ourselves and be explicit about the behavior with `stable` compared to `alpha` 😞  May be keep the `stable` API behavior as is and add changes being introduced with `alpha` in the end of the section 🤔

sbwsg replied:
Yeah you're right it was confusing in this context.  I removed mention of the alpha api from here.

sbwsg reviewed (commented):
I rebased off of latest `main`; the weird artifacts are gone now.

pritidesai commented on pkg/apis/pipeline/v1beta1/task_validation.go:
```diff
@@ -110,6 +112,46 @@ func validateDeclaredWorkspaces(workspaces []WorkspaceDeclaration, steps []Step,
+	for stepIdx, step := range steps {
+		if len(step.Workspaces) != 0 {
+			errs = errs.Also(ValidateEnabledAPIFields(ctx, "step workspaces", config.AlphaAPIFields).ViaIndex(stepIdx).ViaField("steps"))
+		}
```
Should we return an error here or continue validating workspaces even when a feature gate is not set to `alpha`? 🤔 Generally we accumulate all the validation errors before exiting.

sbwsg replied:
Yeah, I've decided to leave them accumulating. We can always return early in future if we see reports that it's confusing this way.

pritidesai commented:
Is this feature supported for a task within a pipeline? if so, will be great to have an example for the same 🙏

sbwsg commented:
> Is this feature supported for a task within a pipeline? if so, will be great to have an example for the same 🙏

Added a pipelinerun example.

pritidesai commented:
> Added a pipelinerun example.

Thanks a bunch @sbwsg, one more question (sorry for nagging on this, trying to make sure I understand it since this is the first alpha feature 😜 ). I am not seeing the examples under `v1beta1/taskruns/alpha` and `v1beta1/pipelineruns/alpha` ran as part of the [integration](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/3700/pull-tekton-pipeline-integration-tests/1389891147064676355/) tests. Is that expected? The examples are more of a reference rather than validating the functionality?

afrittoli commented:
> # Release Notes
> ```
> - Sidecars in a Task now receive Workspaces, just like Steps do.
> - Steps and Sidecars can now specify which Workspaces they want access to. This isolates the workspace to only those Steps or Sidecars that need them, allowing Task authors to limit exposure of sensitive data like credentials.
> - The default behaviour for workspaces in Steps remains the same - all Steps get access to them.
> ```

Could you mention in the release notes that this change in behaviour is alpha / gated by a feature flag - so there will be no change in default behaviour when upgrading to v0.24? 🙏

sbwsg commented:
> > Added a pipelinerun example.
> 
> Thanks a bunch @sbwsg, one more question (sorry for nagging on this, trying to make sure I understand it since this is the first alpha feature 😜 ). I am not seeing the examples under `v1beta1/taskruns/alpha` and `v1beta1/pipelineruns/alpha` ran as part of the [integration](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/3700/pull-tekton-pipeline-integration-tests/1389891147064676355/) tests. Is that expected? The examples are more of a reference rather than validating the functionality?

I created https://github.com/tektoncd/pipeline/issues/3896 to tackle it and will pick it up asap.

> Could you mention in the release notes that this change in behaviour is alpha / gated by a feature flag - so there will be no change in default behaviour when upgrading to v0.24? 🙏

Sure thing!  Done.

afrittoli commented on examples/v1beta1/taskruns/alpha/workspace-isolation.yaml:
```diff
@@ -0,0 +1,55 @@
+          sleep 1
+        done
+        echo "Saw ready file"
+    - name: check-signals-acess
```
NIT: s/acess/access

pritidesai commented on pkg/workspace/apply.go:
```diff
@@ -98,6 +100,23 @@ func Apply(ts v1beta1.TaskSpec, wb []v1beta1.WorkspaceBinding, v map[string]core
 
+	isolatedWorkspaces := sets.NewString()
+
+	alphaAPIEnabled := config.FromContextOrDefaults(ctx).FeatureFlags.EnableAPIFields == "alpha"
```
NIT: can use `config.AlphaAPIFields` instead of string literal `"alpha"`

afrittoli reviewed (commented):
Review in progress 🙏 I will finish tomorrow if no-one gets to it first

pritidesai commented on pkg/workspace/apply.go:
```diff
@@ -106,12 +125,23 @@ func Apply(ts v1beta1.TaskSpec, wb []v1beta1.WorkspaceBinding, v map[string]core
+			if isolatedWorkspaces.Has(w.Name) {
+				mountAsIsolatedWorkspace(ts, w.Name, volumeMount)
+			} else {
+				mountAsSharedWorkspace(ts, volumeMount)
```
I am guessing this will go away once this feature is promoted to stable 🤔

pritidesai replied:
or rather the else ⬇️

pritidesai commented on pkg/workspace/apply_test.go:
```diff
@@ -522,3 +524,310 @@ func TestApply(t *testing.T) {
 	}
 }
+
+func TestApply_IsolatedWorkspaces(t *testing.T) {
```
NIT: The tests included here are pretty comprehensive (go coverage is very obvious) 😍  but will be great to have a test with two different workspaces out of which one is isolated to a step and one is isolated to a sidecar.

pritidesai commented:
thanks a bunch @sbwsg for your patience on this 🙏 

It looks great. I am a little nervous without having `alpha` examples running as part of the CI process. But not a blocker here. I have validated all the examples on my local cluster and works smooth. We can wait until https://github.com/tektoncd/pipeline/issues/3896 is implemented (hopefully soon 🤞). I have also validated [workspaces.yaml](https://github.com/tektoncd/pipeline/blob/main/examples/v1beta1/pipelineruns/workspaces.yaml) `pipelineRun` with `alpha` API and works as expected without any issues.

/lgtm
